### PR TITLE
refactor: introduce `ParOps.parMapValuesSeq`

### LINKED ISSUE #2974: Add `ParOps.mapValues` for `Map[K, V]` and use it broadly
The idea is to have a function `mapValues` which applies a function `V1 => V2` to all values in a Map in parallel and rebuilds the map.

We can then use it in several places in the compiler.

### PR DESCRIPTION
**Pull request by sockmaster27**
Fixes #2974

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/util/ParOps.scala [hunk 1]
```diff
@@ -52,6 +52,18 @@ object ParOps {
       case (k, v) => (k, f(v))
     }.toMap
 
+  /**
+    * Apply the given fallible function `f` to each value in the map `m` in parallel,
+    * returning the resulting map if all calls are successful.
+    */
+  @inline
+  def mapValuesFallible[K, A, B, E](m: Map[K, A])(f: A => Validation[B, E])(implicit flix: Flix): Validation[Map[K, B], E] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/util/ParOps.scala:60]: rename to `parMapSeq`

**Comment by magnus-madsen** [comment]
Has this been applied to every phase where possible?

File: main/src/ca/uwaterloo/flix/util/ParOps.scala (hunk 1 above)
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/util/ParOps.scala:60]: Wouldn't that make it sound like it corresponded to `parMap` instead of `mapValues`?

**Comment by sockmaster27** [comment]
> Has this been applied to every phase where possible?

Yes, I believe so.

File: main/src/ca/uwaterloo/flix/language/phase/Kinder.scala [hunk 2]
```diff
@@ -72,39 +72,29 @@ object Kinder {
...
-            visitEnum(enum, taenv, root).map(sym -> _)
-        }))
-
-        // Extra type annotations are required due to limitations in Scala's type inference.
-        val restrictableEnumsVal = Validation.sequence(ParOps.parMap(root.restrictableEnums)({
-          pair: (Symbol.RestrictableEnumSym, ResolvedAst.Declaration.RestrictableEnum) =>
-            val (sym, enum) = pair
-            visitRestrictableEnum(enum, taenv, root).map(sym -> _)
-        }))
+        val enumsVal = ParOps.mapValuesFallible(root.enums) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Kinder.scala:75]: this  can become `....mapValues(root.enums)(visitEnum(_, taenv, root)`.

Same pattern can be used everywhere else.

File: main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala [hunk 3]
```diff
@@ -59,20 +59,16 @@ object Stratifier {
 
     // Compute the stratification at every datalog expression in the ast.
     val newDefs = flix.subphase("Stratify Defs") {
-      val result = ParOps.parMap(root.defs)(kv => visitDef(kv._2)(root, g, flix).map(d => kv._1 -> d))
-      Validation.sequence(result)
+      ParOps.mapValuesFallible(root.defs)(v => visitDef(v)(root, g, flix))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala:62]: implicit can be omitted? (comment applies everywhere)

Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala:62]: If you mean `(root, g, flix)` then `root` and `g` aren't implicit.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala]: Oh okay.

File: main/src/ca/uwaterloo/flix/util/ParOps.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ParOps.scala:60]: Maybe parMapValuesSeq

Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/util/ParOps.scala:60]: Should `mapValues` then also be renamed to `parMapValues`?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/util/ParOps.scala]: Ah, maybe without the par

Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/util/ParOps.scala]: Actually, looking at the other functions, it looks like they should all just be called `par`-, right?

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Next up https://github.com/flix/flix/issues/6583

**Comment by magnus-madsen** [comment]
I think there are still a few uses of `parMap` which should become `parMapSeq` (if you search for uses of the former). Will you check? Thanks!

**Comment by sockmaster27** [comment]
> I think there are still a few uses of `parMap` which should become `parMapSeq` (if you search for uses of the former). Will you check? Thanks!

Not any that can utilize `parMapValuesSeq`, but should we introduce the more general `parMapSeq`?

**Comment by magnus-madsen** [comment]
Yes, let us do it.

**Pull request merged**